dev/PowerNotifications/PowerNotifications.cpp

```diff
@@ -0,0 +1,501 @@
+                    break;
+
+                case PowerSupplyStatusFn:
+                    break;
+
+                case RemainingChargePercentFn:
+                    break;
+
+                case RemainingDischargeTimeFn:
+                    check_hresult(GetDischargeTime(&m_cachedDischargeTime));
+                    break;
+
+                case PowerSourceStatusFn:
+                    check_hresult(GetPowerCondition(&m_cachedPowerCondition));
+                    break;
+
+                case DisplayStatusFn:
+                    check_hresult(GetDisplayStatus(&m_cachedDisplayStatus));
+                    break;
+
+                case SystemIdleStatusFn:
+                    break;
+
+                case PowerSchemePersonalityFn:
+                    check_hresult(GetPowerSchemePersonality(&m_cachedPowerSchemePersonality));
+                    break;
+
+                case UserPresenceStatusFn:
+                    check_hresult(GetUserPresenceStatus(&m_cachedUserPresenceStatus));
+                    break;
+
+                case SystemAwayModeStatusFn:
+                    check_hresult(GetSystemAwayModeStatus(&m_cachedSystemAwayModeStatus));
+                    break;
+
+                default: throw E_NOTIMPL;
+            }
+        }
+
+    }
+
+    EnergySaverStatus PowerManager::EnergySaverStatus()
+    {
+        CheckRegistrationAndOrUpdateValue(EnergySaverStatusFn);
+        return static_cast<winrt::Microsoft::ProjectReunion::EnergySaverStatus>(m_cachedEnergySaverStatus);
+    }
+
+    EventToken PowerManager::EnergySaverStatusChanged(PowerEventHandle const& handler)
+    {
+        return AddCallback(EnergySaverStatusFn, handler);
+    }
+
+    void PowerManager::EnergySaverStatusChanged(EventToken const& token)
+    {
+        RemoveCallback(EnergySaverStatusFn, token);
+    }
+
+    void PowerManager::EnergySaverStatusChanged_Callback(::EnergySaverStatus energySaverStatus)
+    {
+        m_cachedEnergySaverStatus = energySaverStatus;
+        FireEvent(EnergySaverStatusFn);
+    }
+
+    void PowerManager::ProcessCompositeBatteryStatus(CompositeBatteryStatus const& compositeBatteryStatus)
+    {
+        //Calculate battery charge
+        auto fullChargedCapacity = compositeBatteryStatus.Information.FullChargedCapacity;
+        auto remainingCapacity = compositeBatteryStatus.Status.Capacity;
+        auto newRemainingChargePercent = static_cast<int>((remainingCapacity * 200) / fullChargedCapacity);
+        newRemainingChargePercent += 1;
+        newRemainingChargePercent /= 2;
+        m_batteryChargePercent = newRemainingChargePercent;
+
+        auto powerState = compositeBatteryStatus.Status.PowerState;
+
+        //Set battery status
+        if (compositeBatteryStatus.ActiveBatteryCount == 0)
+        {
+            m_batteryStatus = BatteryStatus::NotPresent;
+        }
+        else if ((powerState & BATTERY_DISCHARGING) != FALSE)
```

Comment:
Try using the WIL helpers `WI_IsFlagSet` / `WI_IsFlagClear` here; I'm having problems parsing the booleans.